src/libostree/ostree-repo-pull.c

```diff
@@ -4027,12 +4027,44 @@ ostree_repo_pull_with_options (OstreeRepo             *self,
     if (pull_data->gpg_verify_summary && bytes_summary && bytes_sig)
       {
         g_autoptr(OstreeGpgVerifyResult) result = NULL;
+        g_autoptr(GError) temp_error = NULL;
 
         result = ostree_repo_verify_summary (self, pull_data->remote_name,
                                              bytes_summary, bytes_sig,
                                              cancellable, error);
-        if (!ostree_gpg_verify_result_require_valid_signature (result, error))
-          goto out;
+        if (!ostree_gpg_verify_result_require_valid_signature (result, &temp_error))
+          {
+            if (summary_from_cache)
+              {
+                /* The cached summary doesn't match, fetch a new one and verify again */
+                summary_from_cache = FALSE;
+                g_clear_pointer (&bytes_summary, (GDestroyNotify)g_bytes_unref);
+                g_warning ("Remote %s cached summary invalid, pulling new version",
+                           pull_data->remote_name);
+                if (!_ostree_fetcher_mirrored_request_to_membuf (pull_data->fetcher,
+                                                                 pull_data->meta_mirrorlist,
+                                                                 "summary",
+                                                                 OSTREE_FETCHER_REQUEST_OPTIONAL_CONTENT,
+                                                                 pull_data->n_network_retries,
+                                                                 &bytes_summary,
+                                                                 OSTREE_MAX_METADATA_SIZE,
+                                                                 cancellable, error))
+                  goto out;
+
+                g_autoptr(OstreeGpgVerifyResult) retry =
+                  ostree_repo_verify_summary (self, pull_data->remote_name,
+                                              bytes_summary, bytes_sig,
+                                              cancellable, error);
+                if (!ostree_gpg_verify_result_require_valid_signature (retry, error))
+                  goto out;
+              }
+            else
+              {
+                if (temp_error != NULL)
```

Comment:
Hmm, if `temp_error` is `NULL` at this point, then there's something wrong. Otherwise, it means there's a chance we could return `FALSE` with no error set at all.

IOW, let's just drop this and let the `g_propagate_error` precondition check yell at us or alternatively make it a `g_assert`?

Reply:
What happens (I believe, need to go back and check) is that the return from `ostree_repo_verify_summary` isn't checked and just passed into `ostree_gpg_verify_result_require_valid_signature`, which immediately returns `FALSE` if the result is `NULL` without setting an error.

I started looking at either checking the result of `ostree_repo_verify_summary`, but it was making the code a little hairy. It would probably work fine if I passed in `&temp_error` to `ostree_repo_verify_summary`, though, since `ostree_gpg_verify_result_require_valid_signature` will just return immediately without setting the error.